tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ClassEmitter.scala

```diff
@@ -276,31 +268,42 @@ private[emitter] final class ClassEmitter(jsGen: JSGen) {
    *
    *  The generated function performs the following steps:
    *
-   *  - Call the super constructor, as specified by `superCtorCall`
    *  - Create the fields of the current class, initialized to the zero of
    *    their respective types
+   *    - In ECMAScript 5.1, all the fields in the parent chain are directly
+   *      created, avoiding expensive super calls
+   *    - In ECMAScript 2015, calling the super constructor is mandatory, so we
+   *      first call it then create the fields directly declared in this class
    *  - Executes the body of the `init` method to inline, if any
    *
    *  @param tree
    *    The `LinkedClass` for which we generate a JS constructor.
    *
-   *  @param superCtorCall
-   *    The call to the super JS constructor. This will be different depending
-   *    on the output mode.
-   *
    *  @param initToInline
    *    The `init` method to inline in the JS constructor, if any.
    */
-  private def genJSConstructorFun(tree: LinkedClass, superCtorCall: js.Tree,
+  private def genJSConstructorFun(tree: LinkedClass,
       initToInline: Option[MethodDef])(
       implicit globalKnowledge: GlobalKnowledge): WithGlobals[js.Function] = {
 
     implicit val pos = tree.pos
 
-    val fieldDefs = genFieldDefsOfScalaClass(tree)
+    val superCtorCallAndFieldDefs = outputMode match {
+      case OutputMode.ECMAScript51Isolated =>
+        val allFields =
+          globalKnowledge.getAllScalaClassFieldDefs(tree.encodedName)
+        genFieldDefsOfScalaClass(tree, allFields)
```

Comment:
Just to double check: This is safe because we never inline the init of a class that has instantiated subclasses? Maybe add a comment about this?

Reply:
Ah, I see what you mean. Yes, indeed, that's right.